Include feedback from OSF (Hypothesis), use CRediT taxonomy for contributions, includes editorial review

nuest opened the pull request.

vsoch commented:
@nuest aren't some of these changes done by @benmarwick ? Why don't we see his commits here, and you are listed as the author (e.g., see blame view https://github.com/nuest/ten-simple-rules-dockerfiles/blame/390f259e41cc5658383bec39e7c9aebf0a52a8f1/ten-simple-rules-dockerfiles.Rmd#L123?)

vsoch commented:
For the reviewer, it would be good to maintain his commits, and then make your changes on top of that (with different commits) otherwise it's impossible to see what you've changed, and the only possible way to review is to re-read everything. Are you able to adjust this PR to properly reflect his contributions, and show your changes cleanly in separate commits?

nuest commented:
@vsoch Maybe you mean @bdevans ? I only did some reformatting on his changes (adding new lines) in https://github.com/nuest/ten-simple-rules-dockerfiles/pull/68/commits/16442f16e861acdbdc142c7b606a204791e7398a

Should I re-do those in a specific commit?

vsoch commented:
oops yes wrong GitHub alias :) yes that's the right b-starting alias I meant to say!

nuest commented:
OK - with the right Ben mentioned - I don't know how what to do differently about the change to @bdevans's additions.

vsoch commented:
@nuest the commit I'm looking for is https://github.com/nuest/ten-simple-rules-dockerfiles/pull/64 - is it still in there? I'm wondering because I'm seeing blocks of text (e.g., the figure caption, which I remember reviewing) that are marked as new here.

nuest commented:
I only changed one word in the figure caption, maybe the online diff does not catch that well because it is a very lon line, so here's a screenshot from my local git tool:

![image](https://user-images.githubusercontent.com/1325054/81431151-f87c3900-9160-11ea-88c8-2eafd7b7b635.png)

------

The PR was merged in https://github.com/nuest/ten-simple-rules-dockerfiles/commit/74c7d2c9a3adcf46fce5e4cce8e98116d992714a

vsoch commented:
Can you walk me through how you merged it (so why your GitHub alias has credit for the commit?) Was it a squash and merge in the interface (and did you add yourself as a co-author?) It's okay if there was a bit of mess up - but I do think it's important that contributors are given proper credit for their work. think we could rectify it here if we add a comment to the merge message "Co-Authored by" and then put @bdevans email. see https://help.github.com/en/github/committing-changes-to-your-project/creating-a-commit-with-multiple-authors.

vsoch commented:
okay I think we are probably okay then - it must be that you changed around that paragraph enough for it to be fully green again. I just chose (another) random line from that PR and @bdevans gets the correct attribution:

![image](https://user-images.githubusercontent.com/814322/81431354-3ad05680-911e-11ea-8d65-c6b97d77ad40.png)

I'll continue with my review, just wanted to be careful about these things! :)

nuest commented:
If you go back one step in the blame, then the commit is by @bdevans : https://github.com/nuest/ten-simple-rules-dockerfiles/blame/93568a7222530ded90e4eca1b0cba65119e33772/ten-simple-rules-dockerfiles.Rmd#L126

nuest commented:
Thanks for being careful! I was just talking to someone about the "tracking contributions" aspect of writing papers on GitHub, and why I'm not expecting it, it _is_ good to know that someone could trace back individual contributions.

vsoch commented on ten-simple-rules-dockerfiles.Rmd:
````diff
@@ -120,10 +120,12 @@ A `Dockerfile` consists of a sequence of instructions to copy files and install
 See Listing&nbsp;1 for a full `Dockerfile`, which we will refer to throughout this article.
 
-```{r container-analogy, echo=FALSE, fig.cap="The workflow to create Docker containers by analogy. Containers begin with a Dockerfile, a recipe for building the computational environment (analogous to source code in a compiled programming language). This is used to build an image with the `docker build` command (analogous to compiling the source code into a binary). Finally the image is used to l ...
+```{r container-analogy, echo=FALSE, fig.cap="The workflow to create Docker containers by analogy. Containers begin with a Dockerfile, a recipe for building the computational environment (analogous to source code in a compiled programming language). This is used to build an image with the `docker build` command (analogous to compiling the source code into an executable binary). Finally the image  ...
````
What is different here (I don't see it).

nuest replied:
Changed "compiling the source code into a binary" to "compiling the source code into an executable binary", because I'm not sure everyone immediately knows what a "binary" is.

Maybe "executable" is even a better term: https://en.wikipedia.org/wiki/Executable

@bdevans Would you be fine with saying "analogous to compiling the source code into an executable file" ?

bdevans replied:
@nuest That would be fine I think. I chose "binary" because I wanted to convey the idea that an image is no longer plain text (like the Dockerfile that generated it) and that it can be run like an executable. 

Maybe "analogous to compiling the source code into an executable (binary) file" would be even better?

nuest replied:
Yes, let's go with both.

nuest commented:
@vsoch Can you wait 10 minutes with your review? Then I'll push some more review-worthy changes.

vsoch commented on ten-simple-rules-dockerfiles.Rmd:
```diff
@@ -440,12 +442,12 @@ A recommended ordering based on these considerations is as follows.
 6. Default commands and entrypoints
 
-# 6. Define version numbers for reproducible builds {-}
+# 6. Specify software versions {-}
```
Much more succinct, I like it.

nuest replied:
Stole that from #65 !

vsoch commented on ten-simple-rules-dockerfiles.Rmd:
```diff
@@ -531,8 +533,8 @@ Unless specific features are needed, bind mounts are also preferable to [storage
-You should avoid installing software packages from source after `COPY`ing the code into the image, because the connection between the file outside of the image and the one copied in is easily lost.
-If you cannot publish a software project in the appropriate language's package structure, you should make sure the files are always published together with the `Dockerfile` (see&nbsp;\ruleref{rule:publish}).
+If you cannot publish a software project in the appropriate language's package structure, you should make sure the source code files you `COPY` into the image are published together with the `Dockerfile` (see&nbsp;\ruleref{rule:publish}).
+If the code and `Dockerfile` are in the same repository, the origin and results of the `COPY` instruction and subsequent installation code are understandable in the future.
```
This is much more clear. COPY isn't bad or evil, unless you don't provide the stuffs copied somewhere :)

nuest commented:
@vsoch What do you think about removing Kitematic because it's depracted?

In https://hyp.is/GJYCroDZEeqMIncvK264wg/osf.io/fsd7t/
![image](https://user-images.githubusercontent.com/1325054/81432533-6164b080-9163-11ea-8291-1bcb53f0cc43.png)

vsoch commented:
I don't use a Mac so I don't (or didn't) use Kinematic, but if it's reported to be deprecated I agree it should be removed.

vsoch commented on ten-simple-rules-dockerfiles.Rmd:
```diff
@@ -420,6 +428,7 @@ _Listing&nbsp;3_: Example labels.
 
 It is often helpful to provide usage instructions, i.e., how to `docker build` and `docker run` the image, _within_ the `Dockerfile`, either at the top or bottom where the reader is likely to find it.
 Such documentation is especially relevant if volume mounts, specific names, or ports are important for using the container, see for example the final lines of Listing&nbsp;1.
+These instructions are not limited to `docker <command>`, but include usage of bespoke scripts, `Makefile`s, or `docker-compose` (see \ruleref{rule:interactive} and \ruleref{rule:usage}).
```
I've sensed that "bespoke" is a popular word :) I think it's overused a bit, maybe take it out here, or choose a different term. I wonder how many times we use it in the paper? Or maybe it just sticks out to me because I didn't use / see it a lot before this paper? :thinking:

nuest replied:
We use it :two: times :-). I admit I like it.

vsoch replied:
ahh okay that's not very many, I'm ok with that :)

vsoch commented on ten-simple-rules-dockerfiles.Rmd:
```diff
@@ -684,10 +697,10 @@ Online collaboration platforms (e.g., GitHub, GitLab) also make it easy to use C
 
 While there are exceptions to the rule (cf.&nbsp;@kim_bio-docklets_2017), it is generally a simple and clear approach to provide one `Dockerfile` per project.
-Alternatively, you could build an _image stack_ to serve different use cases with a common base image, but this reduces understandability and scatters information across multiple places.
+Alternatively, you could build, or even better contribute to existing, _image stacks_ to serve different use cases with a common base image, but this reduces understandability and scatters information across multiple places.
```
Can you explain this more a bit for me? Why would using a popular image stack scatter information ? Don't we advise to use well known / developed bases and this contradicts that?

nuest replied:
The idea is to say "before you write your own stack, see if you can contribute to an existing (community-maintained?) one".

nuest replied:
But that might already be covered in rule 1, which could be a better place for that piece of advice (contribute > build own), IF we choose to leave it in.

vsoch replied:
Ohh I see - I think we should rewrite this then, because I read it as "If you use a stack you will reduce understandability and scatter information across multiple places."

nuest replied:
Well, that's actually what I think. If you (without good reason) use a stack based on 3 or 4 Dockerfiles just to keep these short, then you reduce understandability.

vsoch replied:
Oh so stack == bootstraps? That’s not clear because normally I think of image sets when someone says stack.

vsoch replied:
Here is some better feedback on why this is confusing - when I hear "stack" I think of a stack of containers, not as in bootstrapped from one another, but a group of services. E.g., https://docs.docker.com/engine/swarm/stack-deploy/ - your docker "stack" is your set of containers that are orchestrated together, not one container that has been bootstrapped many times for a final recipe.

vsoch replied:
Maybe instead here you might refer to multiple multi-stage builds, or bootstraps instead of stack?

nuest replied:
Ahh, we indeed understand the term differently. I was thinking of a stack of images that are starting `FROM` one another, see https://github.com/jupyter/docker-stacks and https://www.rocker-project.org/images/, not a _set_ of containers running together.

So I tried to clear this up, and also moved the "collaborate instead of running your own" a bit down within the rule.

vsoch commented on ten-simple-rules-dockerfiles.Rmd:
```diff
@@ -751,6 +764,7 @@ Most importantly, share and exchange your `Dockerfile` freely and collaborate in
 
 DN is supported by the project Opening Reproducible Research II (\href{https://o2r.info/}{https://o2r.info/}; \href{https://www.uni-muenster.de/forschungaz/project/12343}{https://www.uni-muenster.de/forschungaz/project/12343}) funded by the German Research Foundation (DFG) under project number PE 1632/17-1.
 DN and SJE are supported by a Mozilla mini science grant.
+We thank Dav Clark who provided feedback on the preprint [@nust_ten_2020] of this paper.
```
yay!

nuest commented on ten-simple-rules-dockerfiles.Rmd:
```diff
@@ -130,8 +131,10 @@ knitr::include_graphics("container-analogy.png")
 While Docker was the original technology to support the `Dockerfile` format, other container technologies with support include [podman](https://podman.io/)/[buildah](https://github.com/containers/buildah) supported by RedHat, [kaniko](https://github.com/GoogleContainerTools/kaniko), [img](https://github.com/genuinetools/img), and [buildkit](https://github.com/moby/buildkit).
-The Singularity container software [@kurtzer_singularity_2017] is optimised for high performance computing and although it uses its own format, the _Singularity recipe_, it can import and run Docker images.
-Although the Singularity recipe format is different, the rules here are transferable to some extent.
+The Singularity container software [@kurtzer_singularity_2017] is optimised for scientific computing and the security needs of HPC environments.
```
So far we have Britisch spelling, consistently AFAICS, so I suggest to stick to `-ised`.

vsoch replied:
Are you sure? I changed a lot of s's to z's :) Is this something that your spelling / grammar checker tool can check? Do we have more british or more American authors (I might lose this one!)

nuest replied:
Yes, I think we're more :gb: than :us: :-)

I'll double check for consistency.

nuest replied:
https://github.com/nuest/ten-simple-rules-dockerfiles/pull/68/commits/7501c49e8cfb3368384d4b23890ab6ed7e263792

nuest commented on ten-simple-rules-dockerfiles.Rmd:
```diff
@@ -269,17 +272,18 @@ You can identify these images by the missing the user portion of the image name,
 While some organizations can be trusted to update containers with security fixes (see list below), for most individual accounts that provide ready to use images, it is likely that these will not be updated regularly.
 It is even possible that images or `Dockerfile`s may disappear, or that images are published with malicious intent, though we have not heard of any such case in academia.
-Therefore, for security, transparency, and reproducibility, you should only use images where you have access to the `Dockerfile`, and it is suggested to save a copy of the `Dockerfile` in the case that a repository or other distribution goes away.
+Therefore, for security, transparency, and reproducibility, you should only use images where you have access to the `Dockerfile`, e.g., by using automated builds on Docker Hub.
```
I was not aware of the hack and change in practice. Will remove, but I guess mentioning the reason for it goes beyond the scope?

vsoch replied:
There were some smaller ones, but this is the one https://www.zdnet.com/article/docker-hub-hack-exposed-data-of-190000-users/ that at least for me, made me step back and decide to use Quay.io instead, because you can generate repository specific robot tokens. Docker Hub has since added tokens for user login, but that's still a global token that (if discovered) would compromise a user's entire account. So these days I might push containers there directly, but most of the containers I care about are on Quay.io where I have better control of permissions.

I think that particular "Which registry is better" is out of scope for this paper, and it doesn't logically fit here. I think we should just say:

>Therefore, for security, transparency, and reproducibility, you should only use images where you have access to the `Dockerfile`.

It's implied that the images have to come from some registry.

vsoch replied:
And it's not some official, declared change in practice, but (I tended to see) people shirk a bit.

nuest replied:
Thanks for the pointer! Will push an update later, and will have to rethink my whole routine...

nuest commented on ten-simple-rules-dockerfiles.Rmd:
```diff
@@ -309,11 +313,13 @@ This means that you should use indentation, new lines, and comments to make your
 Specifically, carefully indent commands and their arguments to make clear what belongs together, especially when connecting multiple commands in a `RUN` instruction with `&&`.
 Use `\` at the end of a line to break a single command into multiple lines.
 This will ensure that no single line gets too long to comfortably read.
+You may choose to put connecting syntactic elements at the beginning of lines to avoid faulty placement or lingering `&&` as shown in Listing&nbsp;2.
```
Are you OK with Listing 2 having them at the line beginnings (but the other Listings not) ?

vsoch replied:
I think it's okay to show examples that differ, but I wouldn't explicitly say to the user (as we do above) "at the beginning of lines") that one is better than the other. A careful reader would notice our discrepancy in the examples, however, so in this case I'm good with choosing whichever you like better :)

nuest commented on ten-simple-rules-dockerfiles.Rmd:
```diff
@@ -353,17 +359,17 @@ Here is a selection of typical kinds of comments that are useful to include in a
-RUN R -e 'getOption("repos")' && \
-  install2.r \
+RUN R -e 'getOption("repos")' \
+  && install2.r \
```
Extending on https://github.com/nuest/ten-simple-rules-dockerfiles/pull/68#discussion_r422387987, I suggest to leave this example this way, and the others in a way that you and I prefer :-).

Fine with you?

vsoch replied:
yep! Let's just choose one and be consistent and not explicitly state in the text _though shall put the && at the beginning of the line!_

nuest replied:
Well, @davclark suggested this, and I do see the value that his habit reduces the risk of "stray &&" at the end of lines, which happens to me regularly. BUT those will also be caught in the next `docker build` and cause an error, AFAICS.

You are in the "&& at the end of the line" fraction, right?

vsoch replied:
Yes, in favor of readability.

davclark replied:
I think "readability" is pretty subjective here. If you write any Elm, for example, style dictates separators at the beginnings of lines!

But I agree that for the likely mostly Python and Rstats users that will use these recommendations, it's probably more habitual.

As I said in a private note to @nuest, I'm not going to push too hard, as there is little hard data or logic to push us either way. I do make far fewer errors editing Dockerfiles personally if I put separators at the beginning of a line!

vsoch replied:
touche to that, but for me, I make fewer errors at the end :) The comments above by @davclark solidify my assertion that many people will disagree about this and we should not push for either way, but just show one of the examples (whichever @nuest prefers).

nuest commented on ten-simple-rules-dockerfiles.Rmd:
```diff
@@ -398,7 +404,9 @@ LABEL maintainer='D. Nüst <daniel.nuest@uni-muenster.de>' \
+# This ARG instruction adds a build-time variable with the default value 'unspecified'
+# Set it during `docker build` as follows to insert the current date and time into the 
+# label 'org.opencontainers.image.created':
 #   --build-arg BUILD_DATE=$(date -u +"%Y-%m-%dT%H:%M:%SZ")
```
This way it is clear that you can set a default value. Maybe use something else than "undefined", but... `development-version` ?

vsoch replied:
If I were reviewing this Dockerfile, I'd say "if you are building the container, and we always have a reliable way to state the date in the dockerfile, it should not be exposed as a build arg, but always set in the container programatically." So I think the `BUILD_DATE` arg is possibly a bad example. A more appropriate example might be `BUILD_TYPE` or something like that, and then the default would be development:

```
ARG BUILD_TYPE development
```
because if you set `BUILD_DATE` to be an arg and rely on the user to use date() on the command line, it's almost undoubtedly going to be forgotten and the arg will be useless because we'll never know the date.

nuest replied:
I understand. I also thought about this as something that might not be used often, and might be easily forgotten unless automated builds/Makefile etc. are used.

Do you suggest to keep it in?

What would be an example for the non-default value? `BUILD_TYPE=preprint-submission-v1` ? Or would that better be stored in a label?

vsoch replied:
production would be the arg used when the container is deployed /versioned. It would likely drive differences in the container for development vs production.

nuest replied:
I've removed the `.created` label and also the `ARG` example, see further comments in #70

vsoch commented on ten-simple-rules-dockerfiles.Rmd:
```diff
@@ -408,25 +411,56 @@ LABEL maintainer='D. Nüst <daniel.nuest@uni-muenster.de>' \
+## Define versions, parameters, and paths once
+
+The [`ENV` instruction](https://docs.docker.com/engine/reference/builder/#env) in a `Dockerfile` allows for defining environment variables.
+These variables persist inside the container and can be useful for example for (i) setting software versions or paths and reusing them across multiple instructions to avoid mistakes, (ii) specifying metadata intended to be discovered by installed libraries or software, or (ii) adding binaries to the path (`PATH`) or library path (`LD_LIBRARY_PATH`).
```
I think the second (ii) point should be a (iii)

nuest replied:
Fixed.

vsoch commented on ten-simple-rules-dockerfiles.Rmd:
````diff
@@ -408,25 +411,56 @@ LABEL maintainer='D. Nüst <daniel.nuest@uni-muenster.de>' \
+```dockerfile
+# Install UsefulSoft tool in specific version from source
+ENV USEFULSOFT_VERSION 1.0.0
+ENV INSTALL_TEMP=/tmp/useful \
````
Would this really need to be an environment variable (INSTALL_TEMP)

nuest replied:
Nope... just wanted to show how to define multiple ENVs in one `ENV` instruction and needed three variables. Any idea for something more useful?

nuest replied:
Reordered things a little bit.

vsoch commented on ten-simple-rules-dockerfiles.Rmd:
```diff
@@ -408,25 +411,56 @@ LABEL maintainer='D. Nüst <daniel.nuest@uni-muenster.de>' \
+# Install UsefulSoft tool in specific version from source
+ENV USEFULSOFT_VERSION 1.0.0
+ENV INSTALL_TEMP=/tmp/useful \
+  INSTALL_TARGET=/usr/bin
```
Same question here.

vsoch commented on ten-simple-rules-dockerfiles.Rmd:
```diff
@@ -408,25 +411,56 @@ LABEL maintainer='D. Nüst <daniel.nuest@uni-muenster.de>' \
+  rm -r $INSTALL_TEMP $USEFULSOFT_VERSION $USEFULSOFT_VERSION.zip
+
+# Puth UsefulSoft tool on the path for subsequent instructions
+ENV PATH $PATH:$INSTALL_TARGET
```
Isn't /usr/bin usually already on the PATH?

nuest replied:
:facepalm:

nuest replied:
You are so right.

nuest commented:
@vsoch I hope I could improve the text based on your feedback. I think we made some good improvements on clarity, and might even have added a useful new idea or two.

nuest commented:
@vsoch @benmarwick @sje30 @betatim @psychemedia @bdevans

This is an update addressing the comments we had on OSF, consistent use of language, missing formal requirements (see #24), and a change of the contribution section to use the [CRediT Taxonomy](https://casrai.org/credit/).

_Please everyone take a look at the updated contribution section._

Also, the "author summary" is a first attempt that would benefit from one/two second pairs of eyes. Thanks!

@bdevans Please carefully check your affiliation in the Rmd header. Thanks for your contribution! :tada:

bdevans commented on ten-simple-rules-dockerfiles.Rmd:
```diff
@@ -32,15 +33,28 @@ author:
+  - name: Benjamin Evans
+    # https://github.com/bdevans
+    email: benjamin.evans@bristol.ac.uk
+    affiliation: "School of Psychological Science, University of Bristol, GB"
```
For consistency, it should be "School of Psychological Science, University of Bristol, Bristol, GB".

bdevans replied:
Also:
1.  "Stanford University, Stanford, CA, US" should be "Stanford University, Stanford, CA, USA".
2. "University of Washington, Seattle, USA" should be "University of Washington, Seattle, WA, USA" if you include the state or else remove CA from above.

bdevans replied:
My affiliation is otherwise correct, thank you.

nuest replied:
Thanks, well spotted!

bdevans commented on ten-simple-rules-dockerfiles.Rmd:
```diff
@@ -828,6 +833,7 @@ BM contributed to writing – review & editing.
 SJE contributed to conceptualisation and writing – review & editing.
 THe contributed to conceptualisation.
 THi contributed to writing – review & editing.
+BE contributed to visulisation and writing – review & editing.
```
Thanks for my inclusion!

Would revising the overarching structure of the rules through the summary (figure) count as contributing to conceptualisation too?

nuest replied:
I'm not sure, but would tend to say not yet.

_Would you be so kind to raise this with the other authors, maybe via email?_ Not everyone has double-checked their new contributions statement anyway, so that would be helpful to trigger them looking and pinching in.

vsoch replied:
I would tend to say no, conceptualization implies putting ideas together before anything exists, and creating the original thing. You’d have to time travel :)

vsoch replied:
I think for me it would make sense to add reviewing and editing - I’ve been doing a lot of that too over this last week!

nuest replied:
Will add that.

bdevans replied:
Sorry just getting to this. The CRediT role is defined as: 
> "Conceptualization – Ideas; formulation or evolution of overarching research goals and aims." 

This includes "evolution" which, unlike "formulation" would not require time travel ;) My thinking was that shaping/restructuring the rules aligns well with this, as that is *the* concept of the paper but happy to accept the consensus view. Should I still email about this?

vsoch commented on Makefile:
```diff
@@ -1,4 +1,10 @@
 	Rscript -e 'rmarkdown::render("$<")'
 
+edit:
+	PASSWORD=simple
```
Maybe this should not be in plain sight, even if it's just for development?

nuest replied:
I think we don't ever suggest to run the container other than somewhere locally, and I would not even expect someone other than the authors to do so, so I'd say we leave that it. More important things to work on ;-)

vsoch commented on ten-simple-rules-dockerfiles.Rmd:
```diff
@@ -440,30 +440,29 @@ _Listing&nbsp;3_: Commonly used labels.
-ENV INSTALL_TEMP=/tmp/useful \
-  INSTALL_TARGET=/usr/bin
-  
+ENV USEFULSOFT_VERSION=1.0.0 \
```
Does /worksapce exist? I don't see a mkdir -p or WORKDIR and then we use the bin as a target. Unless the install.sh script creates all required parent folders it might result in an error.

nuest replied:
Listing 4 doesn't start with a `FROM` statement, so I suggest to make that clear in the Listing description that this is a partial Dockerfile. Would that fix this?

vsoch replied:
Ah, maybe it would be more clear to have ... at the top, so it's clearly not the entire file. Eg..,

```
...
ENV pancakes aregreat
RUN apt-get update && apt-get install -y breakfast
```

vsoch commented on ten-simple-rules-dockerfiles.Rmd:
```diff
@@ -440,30 +440,29 @@ _Listing&nbsp;3_: Commonly used labels.
+  bash install.sh --target $USEFULSOFT_INSTALLDIR && \
   cd .. && \
-  rm -r $INSTALL_TEMP $USEFULSOFT_VERSION $USEFULSOFT_VERSION.zip
+  rm -r $USEFULSOFT_VERSION $USEFULSOFT_VERSION.zip
```
should there be a space between these?

nuest replied:
One is the directory, one is the .zip-file. Is this better?

```
RUN wget http://download-server.com/useful_software/$USEFULSOFT_VERSION/useful-$USEFULSOFT_VERSION.zip && \
  unzip useful-$USEFULSOFT_VERSION.zip -d useful-src && \
  cd useful-src && \
  bash install.sh --target $USEFULSOFT_INSTALLDIR && \
  cd .. && \
  rm -r useful-src useful-$USEFULSOFT_VERSION.zip
```

vsoch replied:
Ahh okay! My mistake, I thought you were putting together a base directory and zip file.

nuest commented:
@vsoch I' like to merge this to avoid precisely looking at previous changes. I'll resolve your final comments asap, let me know when you're done.

https://github.com/nuest/ten-simple-rules-dockerfiles/pull/65#discussion_r420388833 also points to you and @bdevans thinking there's content that could be cut, so we should get `master` into a good state.
